File: Java-Core/src/main/java/presto/grammar/CodeArgument.java
package presto.grammar;

import presto.error.SyntaxError;
import presto.parser.Dialect;
import presto.runtime.Context;
import presto.type.CodeType;
import presto.type.IType;
import presto.utils.CodeWriter;
import presto.utils.Utils;

public class CodeArgument extends BaseArgument implements ITypedArgument {
	
	public CodeArgument(Identifier name) {
		super(name);
	}
	
	@Override
	public IType getType() {
		return CodeType.instance();
	}
	
	@Override
	public String getSignature(Dialect dialect) {
		return CodeType.instance().getName() + " " + name;
	}
	
	@Override
	public String getProto(Context context) {
		return CodeType.instance().getName().toString();
	}
	
	@Override
	public void toDialect(CodeWriter writer) {
		writer.append(CodeType.instance().getName().toString());
		writer.append(" ");
		writer.append(name);
	}
	
	@Override
	public String toString() {
		return name.toString() + ':' + CodeType.instance().getName();
	}
	
	@Override
	public boolean equals(Object obj) {
		if(obj==this)
			return true;
		if(obj==null)
			return false;
		if(!(obj instanceof CodeArgument))
			return false;
		CodeArgument other = (CodeArgument)obj;
		return Utils.equal(this.getIdentifier(),other.getIdentifier());
	}

	@Override
	public void register(Context context) throws SyntaxError {
		INamed actual = context.getRegisteredValue(INamed.class, name);
		if(actual!=null)
			throw new SyntaxError("Duplicate argument: \"" + name + "\"");
		context.registerValue(this);
	}
	
	@Override
	public void check(Context context) throws SyntaxError {
	}
	
	@Override
	public IType getType(Context context) throws SyntaxError {
		return CodeType.instance();
	}

}

File: Java-CodeServer/src/main/resources/js/prompto/grammar/CodeArgument.js
var SyntaxError = require("../error/SyntaxError").SyntaxError;
var CodeType = require("../type/CodeType").CodeType;
var Argument = require("./Argument").Argument;

function CodeArgument(name) {
	Argument.call(this, name);
	return this;
}

CodeArgument.prototype = Object.create(Argument.prototype);
CodeArgument.prototype.constructor = CodeArgument;

/*
	
@Override
public IType getType() {
	return CodeType.instance();
}

@Override
public String getSignature(Dialect dialect) {
	return name + ':' + CodeType.instance().getName();
}
*/

CodeArgument.prototype.getProto = function(context) {
	return CodeType.instance.name;
};

/*
@Override
public String toString() {
	return name + ':' + CodeType.instance().getName();
}

@Override
public boolean equals(Object obj) {
	if(obj==this)
		return true;
	if(obj==null)
		return false;
	if(!(obj instanceof CodeArgument))
		return false;
	CodeArgument other = (CodeArgument)obj;
	return Utils.equal(this.getName(),other.getName());
}
*/

CodeArgument.prototype.register = function(context) {
	var actual = context.getRegisteredValue(this.name);
	if(actual!=null) {
		throw new SyntaxError("Duplicate argument: \"" + this.name + "\"");
	}
	context.registerValue(this);
};

CodeArgument.prototype.check = function(context) {
	// ok
};

CodeArgument.prototype.getType = function(context) {
	return CodeType.instance;
};

CodeArgument.prototype.toDialect = function(writer) {
    writer.append(CodeType.instance.name);
    writer.append(" ");
    writer.append(this.name);
};


exports.CodeArgument = CodeArgument;